File: pocs/code-review-demo/agents/lead_agent.py
import asyncio
from beeai_framework.agents.experimental import RequirementAgent
from beeai_framework.memory import UnconstrainedMemory
from beeai_framework.tools.think import ThinkTool
from core.backend import get_llm
from core.models import ReviewEvent


def create_lead_agent() -> RequirementAgent:
    return RequirementAgent(
        llm=get_llm(),
        memory=UnconstrainedMemory(),
        tools=[ThinkTool()],
        role="lead_reviewer",
        instructions=(
            "You are the lead engineer consolidating a code review. "
            "You will receive findings from the dev, security, and performance agents. "
            "Summarize all findings and issue a final verdict: APPROVED, CHANGES_REQUESTED, or BLOCKED. "
            "BLOCKED is required if any critical security issue was found."
        ),
    )


async def run_lead_with_approval(
    lead_agent,
    findings: str,
    ws_send,
    approval_event: asyncio.Event,
) -> str:
    await ws_send(
        ReviewEvent(
            type="human_approval",
            agent="lead",
            message="Review complete. Approve to finalize verdict?",
            severity=None,
        ).to_json()
    )

    await approval_event.wait()

    result = await lead_agent.run(findings)
    return result.output.text

File: pocs/code-review-demo-ts/src/agents/leadAgent.ts
import { RequirementAgent } from "beeai-framework/agents/requirement/agent";
import { ConditionalRequirement } from "beeai-framework/agents/requirement/requirements/conditional";
import { ThinkTool } from "beeai-framework/tools/think";
import { getLLM } from "../core/backend.js";

export async function createLeadAgent() {
  return new RequirementAgent({
    llm: await getLLM(),
    tools: [new ThinkTool()],
    instructions: `You are the lead code reviewer. You will receive consolidated findings from Dev, Security, and Performance agents.
Carefully review all findings and decide on one of: APPROVED | CHANGES_REQUESTED | BLOCKED.
Rules:
- A single CRITICAL security finding → BLOCKED
- Multiple WARNINGs with no CRITICAL → CHANGES_REQUESTED
- Only INFO findings → APPROVED
Always think before deciding. End your response with a line: VERDICT: <decision>`,
    requirements: [
      new ConditionalRequirement(ThinkTool, { forceAtStep: 1 }),
    ],
  });
}
